rfcs/002-merge-types-terms-syntax.md

```diff
@@ -0,0 +1,452 @@
+   that is a type in the current syntax as a term in the new syntax must behave
+   the same as the corresponding contract.
+
+### `term`
+
+The `term` translation mainly inserts implicit `§` automatically.
+
+- For most term-only constructs (that is everything excepted
+  records and constructors coming from types), `term` is an homomorphism (e.g.
+  for lists `term [t1, ..., tn] = [ term(t1), ..., term(tn)]`).
+- For type-only constructs (ground types, `->`, and `{_: T}`), it is
+  `term(T) = §(type(T))`, that is the contract associated to the type.
+
+#### Records
+
+It remains to treat the case of records, that can mix previously type and term
+constructs.
+
+- a row declaration `foo : Num` is translated as it is. This is actually already
+  supported since the piecewise signature feature. `{foo : Num}` is just
+  equivalent to `{foo | Num}` operationally when `foo` doesn't have a
+  definition.
+- For a tail, because of the predicate enforced in the grammar, we know that a
+  record `{ body; tail}` in `UniTerm` must satisfy `is_record_type({body})`. We
+  just convert it to a type, and derive the corresponding contract: `term({body;
+  tail}) = §type({body; tail})`.
+
+### `type`
+
+Dually to terms:
+
+- For type-only constructs, `type` is an homomorphism. For example,
+  `type(t -> t') = type(t) -> type(t')`.
+- For term-only constructs, we take the opaque type associated to a contract.
+  That is, `type(t) = #term(t)`. For example,
+  `type(fun x => body) = #term(fun x => body)`.
+
+#### Records
+
+As for terms, the case of records remains. For records that are actually the
+embedding of a `Type`, such as `{foo : Num, bar: {baz: Num}}`, the translation
+is obvious. For records that are the embedding of a `Term` without annotations,
+we can simply proceed as above: `type({foo = 1, bar = "baz"}) = #{foo = 1, bar =
+"baz"}`.
+
+The case for records that mixes the both `Term` and `Type` is less clear. Take
+for example: `x : {foo : Num = 1}`. Should we consider this as a totally opaque
+record, that is `type({foo : Num = 1}) = #{foo : Num = 1}` or take advantage
+from the information that `foo` is necessarily a number and set `type({foo : Num
+= 1}) = {foo: Num}`? The problem with the latter approach is that we loose the
+property of *blame safety*, which is that "well-typed program can't be blamed".
+Indeed, the extracted type is less precise that the actual contract. For
+example, `{foo = 2} : {foo : Num, foo = 1}` would pass typechecking but fails at
+runtime on the failure of merging `1` and `2`. There are also multiple
+approximation possible: we could create an opaque type encoding the tuple `(Num,
+Eq 2)`, we could set `foo : Dyn`, etc.
+
+Because of blame safety, and the different possible trade-offs, **we propose to only translate as
+types records that are already record types in the current syntax**. That is:
+
+```
+type({f1: t1, .., fn : tn}) = {f1: type(t1), .., fn: type(tn)}
```

Comment:
This is discussed in the `Record fields` section. This makes sense, and wouldn't have any fundamental problem.

But right now it turns out `§{foo : Num}` and `{foo | Num}` are behaving differently as contracts. The former is more eager, in that `{} | {foo: Num}` will fail immediately, while `{} | #{foo | Num}` will have a `foo` field without definition, that needs to be projected before the contract fails. I imagine this comes originally from the idea of contracts being based on merging. The laziness of `|` allows to specify data incrementally, without having the invalid intermediate values to break the contract: `{foo | #{bar | Num, baz | Str}} & {foo.bar = 1} & {foo.baz = "ha"}` will work, but `{foo | {bar : Num, baz : Str}} & {foo.bar = 1} & {foo.baz = "ha"}` won't because the intermediate value `{foo.bar = 1}` is missing the `baz` field.

I'm not sure this was a totally conscious design decision, or rather that contract-as-a-term-record and contract-from-a-record-type were implemented at different times with different flavors. This surely needs to be made uniform, but here I just wanted to avoid having to make the eager vs lazy decision in this RFC and rather postpone it, as long as the choice is forward-compatible and is not a regression with respect to the current situation (currently, `foo | #{bar | Num}` makes the type of `foo` to be totally opaque, without meaningful static information)

Reply:
Well, to be fair, the choice is not totally forward-compatible, if we choose to either make `#{foo | Num}` or `{foo : Num}` lazy: the semantics of contracts written in on type or the other would be altered. What I meant is that we can make this choice later, and then upgrade the `type()` translation in a forward-compatible manner. Let's say the choice eager vs lazy commute with the change `type({foo | Num}) = #{foo | Num})` => `type({foo | Num}) = {foo : Num}`.